From bba90253637c9fdb295f0068f275092353b13b09 Mon Sep 17 00:00:00 2001
From: Johannes Demel <jdemel@gnuradio.org>
Date: Sat, 5 Jul 2025 16:37:36 +0200
Subject: [PATCH 05/12] ci: Add Werror flag to CI for C compilation

We only used the Werror flag for C++. If we use the Werror flag, we
should mostly use it for our core.

Signed-off-by: Johannes Demel <jdemel@gnuradio.org>
---
 .github/workflows/run-tests.yml |  8 ++++----
 CMakeLists.txt                  | 13 +++++++++++--
 2 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/.github/workflows/run-tests.yml b/.github/workflows/run-tests.yml
index 7e87e35..c4496a3 100644
--- a/.github/workflows/run-tests.yml
+++ b/.github/workflows/run-tests.yml
@@ -51,7 +51,7 @@ jobs:
         env:
           CC: ${{ matrix.compiler.cc }}
           CXX: ${{ matrix.compiler.cxx }}
-        run: mkdir build && cd build && cmake -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ..
+        run: mkdir build && cd build && cmake -DCMAKE_C_FLAGS="-Werror" -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ..
       - name: Build
         run: |
           echo "Build with $(nproc) thread(s)"
@@ -135,7 +135,7 @@ jobs:
           run: |
             cd /volk
             cd build
-            cmake -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ${{ matrix.cmakeargs }} ..
+            cmake -DCMAKE_C_FLAGS="-Werror" -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ${{ matrix.cmakeargs }} ..
             echo "Build with $(nproc) thread(s)"
             make -j$(nproc)
             if [ -f ./cpu_features/list_cpu_features ]; then
@@ -159,7 +159,7 @@ jobs:
       - name: dependencies
         run: sudo apt install python3-mako liborc-dev libgtest-dev libfmt-dev
       - name: configure
-        run: mkdir build && cd build && cmake -DENABLE_STATIC_LIBS=True -DBUILD_EXECUTABLE=ON ..
+        run: mkdir build && cd build && cmake -DENABLE_STATIC_LIBS=True -DCMAKE_C_FLAGS="-Werror" -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ..
       - name: build
         run: cmake --build build -j$(nproc)
       - name: Print info
@@ -236,7 +236,7 @@ jobs:
       - name: dependencies
         run: pip3 install --break-system-packages mako && brew install orc
       - name: configure
-        run: mkdir build && cd build && cmake -DBUILD_EXECUTABLE=ON ..
+        run: mkdir build && cd build && cmake -DCMAKE_C_FLAGS="-Werror" -DCMAKE_CXX_FLAGS="-Werror" -DBUILD_EXECUTABLE=ON ..
       - name: build
         run: cmake --build build --config Debug -j4
       - name: Print info
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 718c9fa..b4d5f21 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -69,7 +69,7 @@ if(CMAKE_C_COMPILER_ID MATCHES "Clang|GNU")
 endif()
 
 list(INSERT CMAKE_MODULE_PATH 0 ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules
-)#location for custom "Modules"
+) #location for custom "Modules"
 
 include(VolkBuildTypes)
 #select the release build type by default to get optimization flags
@@ -114,7 +114,7 @@ set(CROSSCOMPILE_MULTILIB
 
 if(MSVC)
     add_definitions(-D_USE_MATH_DEFINES
-    )#enables math constants on all supported versions of MSVC
+    ) #enables math constants on all supported versions of MSVC
     add_compile_options(/W1) #reduce warnings
     add_compile_options(/wo4309)
     add_compile_options(/wd4752)
@@ -155,11 +155,20 @@ if(VOLK_CPU_FEATURES)
         set(CMAKE_POSITION_INDEPENDENT_CODE
             ON
             CACHE BOOL "Build cpu_features with Position Independent Code (PIC)." FORCE)
+        include(CheckCCompilerFlag)
+        set(CMAKE_C_FLAGS_SAVED "${CMAKE_C_FLAGS}")
+        check_c_compiler_flag(-Wno-unused-function HAVE_WNO_UNUSED_FUNCTION_CMD_LINE_ARG)
+        if(HAVE_WNO_UNUSED_FUNCTION_CMD_LINE_ARG)
+            set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-function")
+        endif(HAVE_WNO_UNUSED_FUNCTION_CMD_LINE_ARG)
+
+        set(BUILD_SHARED_LIBS OFF)
         set(BUILD_SHARED_LIBS_SAVED "${BUILD_SHARED_LIBS}")
         set(BUILD_SHARED_LIBS OFF)
         set(ENABLE_INSTALL OFF)
         add_subdirectory(cpu_features)
         set(BUILD_SHARED_LIBS "${BUILD_SHARED_LIBS_SAVED}")
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS_SAVED}")
     endif()
 else()
     message(STATUS "Building Volk without cpu_features")
-- 
2.47.2

